### PR TITLE
Fix aarch detection

### LINKED ISSUE #266: Python 2.4.0 + aarch compiled without numpy support
~Possibly relevant:~ We encountered a "PyTorch was compiled without NumPy support" error when running on Linux aarch64 + CUDA (on NVIDIA GH200) using the conda-forge build of PyTorch 2.4.0.

Relevant output from `conda list` for the environment in which the error was encountered:

```text
pytorch                   2.4.0           cuda120_py312haadfe8f_200    conda-forge
pytorch-gpu               2.4.0           cuda120py312hecaec72_200    conda-forge
```

Rolling back to 2.3.0 remedied this issue. Looking at the build number, it seems that the build we installed preceded merging PR #252.

_Originally posted by @jcwomack in https://github.com/conda-forge/pytorch-cpu-feedstock/issues/254#issuecomment-2319090902_

### PR DESCRIPTION
**Pull request by hmaarrfk**
Closes #266 
Documenting a few things I learned:

```
find_package(Python REQUIRED COMPONENTS ...
find_package(Python 3 REQUIRED COMPONENTS ...
```

Seem to behave differently.

They seem to be using some custom cmake shims that define the Python_EXECUTABLE variable

```
        CMake.defines(
            args,
            Python_EXECUTABLE=sys.executable,
            TORCH_BUILD_VERSION=version,
            **build_options,
        )
```
https://github.com/pytorch/pytorch/blob/main/tools/setup_helpers/cmake.py#L310

### CODE REVIEW

File: recipe/build.sh [hunk 1]
```diff
@@ -6,6 +6,11 @@ set -ex
 # https://github.com/pytorch/pytorch/blob/v2.3.1/setup.py#L341
 export PACKAGE_TYPE=conda
 
+
+# hmaarrfk - Designed for Pytorch 2.4.0
+# Help detect failures in Numpy Early
+sed -i.bak 's/Python COMPONENTS Interpreter OPTIONAL_COMPONENTS/Python COMPONENTS Interpreter/g' cmake/Dependencies.cmake
```
[the remark this answers is not in the document]
Comment (hmaarrfk) [reply recipe/build.sh:12]: `USE_NUMPY` is still a "hint". the default is "true"

[the remark this answers is not in the document]
Comment (jakirkham) [reply recipe/build.sh:12]: Ah sorry. Misread 🤦‍♂️

**Comment by hmaarrfk** [comment]
I don't understand why it isn't searching for `Python.h` in `${PREFIX}/include/Python3.12` where it is located.

**Comment by h-vetinari** [comment]
Copying out the logs and reinstating some variables for better legibility
```
    find_path considered the following locations:

      $PREFIX/lib/python3.12/site-packages/include/include/python3.12.5/Python.h
      $PREFIX/lib/python3.12/site-packages/include/include/Python.h
      $PREFIX/lib/python3.12/site-packages/include/Python.h
      $PREFIX/lib/python3.12/site-packages/include/python3.12.5/Python.h
      $PREFIX/lib/python3.12/site-packages/include/Python.h
      $PREFIX/lib/python3.12/site-packages/Python.h
      $PREFIX/include/include/python3.12.5/Python.h
      $PREFIX/include/include/Python.h
      $PREFIX/include/Python.h
      $PREFIX/include/python3.12.5/Python.h
      $PREFIX/include/Python.h
      $PREFIX/Python.h
      $PREFIX/$BUILD_PREFIX/include/python3.12.5/Python.h
      $PREFIX/$BUILD_PREFIX/include/Python.h
      $PREFIX/$BUILD_PREFIX/Python.h
      $BUILD_PREFIX/aarch64-conda-linux-gnu/sysroot/$PREFIX/lib/python3.12/site-packages/include/include/python3.12.5/Python.h
      $BUILD_PREFIX/aarch64-conda-linux-gnu/sysroot/$PREFIX/lib/python3.12/site-packages/include/include/Python.h
      $BUILD_PREFIX/aarch64-conda-linux-gnu/sysroot/$PREFIX/lib/python3.12/site-packages/include/Python.h
      $BUILD_PREFIX/aarch64-conda-linux-gnu/sysroot/$PREFIX/lib/python3.12/site-packages/include/python3.12.5/Python.h
      $BUILD_PREFIX/aarch64-conda-linux-gnu/sysroot/$PREFIX/lib/python3.12/site-packages/include/Python.h
      $BUILD_PREFIX/aarch64-conda-linux-gnu/sysroot/$PREFIX/lib/python3.12/site-packages/Python.h
      $BUILD_PREFIX/aarch64-conda-linux-gnu/sysroot/$PREFIX/include/include/python3.12.5/Python.h
      $BUILD_PREFIX/aarch64-conda-linux-gnu/sysroot/$PREFIX/include/include/Python.h
      $BUILD_PREFIX/aarch64-conda-linux-gnu/sysroot/$PREFIX/include/Python.h
      $BUILD_PREFIX/aarch64-conda-linux-gnu/sysroot/$PREFIX/include/python3.12.5/Python.h
      $BUILD_PREFIX/aarch64-conda-linux-gnu/sysroot/$PREFIX/include/Python.h
      $BUILD_PREFIX/aarch64-conda-linux-gnu/sysroot/$PREFIX/Python.h
      $BUILD_PREFIX/aarch64-conda-linux-gnu/sysroot/$BUILD_PREFIX/include/python3.12.5/Python.h
      $BUILD_PREFIX/aarch64-conda-linux-gnu/sysroot/$BUILD_PREFIX/include/Python.h
      $BUILD_PREFIX/aarch64-conda-linux-gnu/sysroot/$BUILD_PREFIX/Python.h
      $SRC_DIR/$PREFIX/lib/python3.12/site-packages/include/include/python3.12.5/Python.h
      $SRC_DIR/$PREFIX/lib/python3.12/site-packages/include/include/Python.h
      $SRC_DIR/$PREFIX/lib/python3.12/site-packages/include/Python.h
      $SRC_DIR/$PREFIX/lib/python3.12/site-packages/include/python3.12.5/Python.h
      $SRC_DIR/$PREFIX/lib/python3.12/site-packages/include/Python.h
      $SRC_DIR/$PREFIX/lib/python3.12/site-packages/Python.h
      $SRC_DIR/$PREFIX/include/include/python3.12.5/Python.h
      $SRC_DIR/$PREFIX/include/include/Python.h
      $SRC_DIR/$PREFIX/include/Python.h
      $SRC_DIR/$PREFIX/include/python3.12.5/Python.h
      $SRC_DIR/$PREFIX/include/Python.h
      $SRC_DIR/$PREFIX/Python.h
      $SRC_DIR/$BUILD_PREFIX/include/python3.12.5/Python.h
      $SRC_DIR/$BUILD_PREFIX/include/Python.h
      $SRC_DIR/$BUILD_PREFIX/Python.h

    The item was not found.
```

the reason is that the variation we need is missing from the pattern ❌  
```
[...]/include/python3.12.5/Python.h     ✅
[...]/include/python3.12/Python.h       ❌
[...]/include/Python.h                  ✅
[...]/Python.h                          ✅
```

**Comment by hmaarrfk** [comment]
The host environments differences are:
image

The build environment differences are
image

**Comment by h-vetinari** [comment]
Ah, the old python2 vs python3 thing in CMake - great catch @isuruf!

BTW, while checking something else, I just saw that this is using a really old conda-build due to:
https://github.com/conda-forge/pytorch-cpu-feedstock/blob/6dd85b3f85a72371b2d3ccf6a386de67e61d667e/conda-forge.yml#L28-L30

File: conda-forge.yml [hunk 2]
```diff
@@ -24,8 +24,5 @@ os_version:
   linux_64: cos7
 provider:
   linux_64: github_actions
-  linux_aarch64: azure
-remote_ci_setup:
-- py-lief=0.12.3
-- conda-forge-ci-setup=3
+  linux_aarch64: azur
```
Comment (hmaarrfk) [inline conda-forge.yml:27]:
```suggestion
  linux_aarch64: azure
```

**Comment by hmaarrfk** [comment]
@conda-forge-admin please rerender

File: recipe/build.sh [hunk 3]
```diff
@@ -6,6 +6,13 @@ set -ex
 # https://github.com/pytorch/pytorch/blob/v2.3.1/setup.py#L341
 export PACKAGE_TYPE=conda
 
+export CMAKE_FIND_DEBUG_MODE=1
```
Comment (hmaarrfk) [inline recipe/build.sh:9]:
```suggestion
# export CMAKE_FIND_DEBUG_MODE=1
```

**Comment by hmaarrfk** [comment]
Now this is strange. 

The build passes. But the tests are failing….

**Comment by hmaarrfk** [comment]
I made the cmake tests stricter than they used to be, things should fail on the CIs now, but I tried to fix it locally with no luck. Even downgrading to cmake 3.29 didn't have the desired effect.

**Comment by hmaarrfk** [comment]
I'm going to try to run the CUDNN9 migration locally with emulation. will take for ever, but maybe its faster than trying to solve this problem.

However, even build 0 does not detect numpy correctly.

```
--   USE_NUMPY             : OFF
```

can be seen in the build logs.

**Comment by hmaarrfk** [comment]
locally i can find numpy when i  use enumation. looking forward to the 60 hour  build...

**Comment by hmaarrfk** [comment]
@conda-forge-admin please rerender

**Comment by hmaarrfk** [comment]
should we move to azure? the failure happens pretty early on.

**Comment by isuruf** [comment]
sure

**Comment by hmaarrfk** [comment]
@conda-forge-admin please rerender

**Comment by hmaarrfk** [comment]
@conda-forge-admin please rerender

**Comment by hmaarrfk** [comment]
:( 


> 2024-09-21T20:45:42.4880641Z   /home/conda/feedstock_root/build_artifacts/libtorch_1726938996574/_build_env/bin/aarch64-conda-linux-gnu-cc -shared -Wl,-O2 -Wl,--sort-common -Wl,--as-needed -Wl,-z,relro -Wl,-z,now -Wl,--allow-shlib-undefined -Wl,-rpath,/home/conda/feedstock_root/build_artifacts/libtorch_1726938996574/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/lib -Wl,-rpath-link,/home/conda/feedstock_root/build_artifacts/libtorch_1726938996574/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/lib -L/home/conda/feedstock_root/build_artifacts/libtorch_1726938996574/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/lib -Wl,-O2 -Wl,--sort-common -Wl,--as-needed -Wl,-z,relro -Wl,-z,now -Wl,--allow-shlib-undefined -Wl,-rpath,/home/conda/feedstock_root/build_artifacts/libtorch_1726938996574/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/lib -Wl,-rpath-link,/home/conda/feedstock_root/build_artifacts/libtorch_1726938996574/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/lib -L/home/conda/feedstock_root/build_artifacts/libtorch_1726938996574/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/lib -Wl,-O2 -Wl,--sort-common -Wl,-z,relro -Wl,-z,lazy -Wl,--allow-shlib-undefined -Wl,-rpath,/home/conda/feedstock_root/build_artifacts/libtorch_1726938996574/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/lib -Wl,-rpath-link,/home/conda/feedstock_root/build_artifacts/libtorch_1726938996574/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/lib -L/home/conda/feedstock_root/build_artifacts/libtorch_1726938996574/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/lib -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O3 -pipe -isystem /home/conda/feedstock_root/build_artifacts/libtorch_1726938996574/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/include -fdebug-prefix-map=/home/conda/feedstock_root/build_artifacts/libtorch_1726938996574/work=/usr/local/src/conda/libtorch-2.4.0 -fdebug-prefix-map=/home/conda/feedstock_root/build_artifacts/libtorch_1726938996574/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl=/usr/local/src/conda-prefix -Wno-deprecated-declarations -Wno-error=maybe-uninitialized -DNDEBUG -D_FORTIFY_SOURCE=2 -O2 -isystem /home/conda/feedstock_root/build_artifacts/libtorch_1726938996574/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/include build/temp.linux-aarch64-cpython-312/torch/csrc/stub.o -L/home/conda/feedstock_root/build_artifacts/libtorch_1726938996574/work/torch/lib -ltorch_python -o build/lib.linux-aarch64-cpython-312/torch/_C.cpython-312-aarch64-linux-gnu.so -Wl,-rpath,$ORIGIN/lib
2024-09-21T20:45:42.4953479Z   /home/conda/feedstock_root/build_artifacts/libtorch_1726938996574/_build_env/bin/../lib/gcc/aarch64-conda-linux-gnu/13.3.0/../../../../aarch64-conda-linux-gnu/bin/ld: cannot find -ltorch_python: No such file or directory

**Comment by Tobias-Fischer** [comment]
Do we need to manually move the `libtorch_python.so` to the correct directory? It lives in `_build_env/venv` currently which doesn't seem correct?
```
2024-09-21T20:45:36.2059664Z   -- Set non-toolchain portion of runtime path of "/home/conda/feedstock_root/build_artifacts/libtorch_1726938996574/_build_env/venv/lib/libc10.so" to "$ORIGIN"
2024-09-21T20:45:38.2142982Z   -- Set non-toolchain portion of runtime path of "/home/conda/feedstock_root/build_artifacts/libtorch_1726938996574/_build_env/venv/lib/libshm.so" to "$ORIGIN"
2024-09-21T20:45:38.2173476Z   -- Set non-toolchain portion of runtime path of "/home/conda/feedstock_root/build_artifacts/libtorch_1726938996574/_build_env/venv/bin/torch_shm_manager" to "$ORIGIN/../lib"
2024-09-21T20:45:38.2504164Z   -- Set non-toolchain portion of runtime path of "/home/conda/feedstock_root/build_artifacts/libtorch_1726938996574/_build_env/venv/lib/libtorch_python.so" to "$ORIGIN"
2024-09-21T20:45:38.5043675Z   -- Set non-toolchain portion of runtime path of "/home/conda/feedstock_root/build_artifacts/libtorch_1726938996574/_build_env/venv/lib/libtorch_cpu.so" to "$ORIGIN"
2024-09-21T20:45:38.5054215Z   -- Set non-toolchain portion of runtime path of "/home/conda/feedstock_root/build_artifacts/libtorch_1726938996574/_build_env/venv/lib/libtorch.so" to "$ORIGIN"
2024-09-21T20:45:38.5079700Z   -- Set non-toolchain portion of runtime path of "/home/conda/feedstock_root/build_artifacts/libtorch_1726938996574/_build_env/venv/lib/libtorch_global_deps.so" to "$ORIGIN"
2024-09-21T20:45:38.5617385Z   -- Set non-toolchain portion of runtime path of "/home/conda/feedstock_root/build_artifacts/libtorch_1726938996574/work/functorch/functorch.so" to "$ORIGIN/../torch/lib"
```

**Comment by hmaarrfk** [comment]
Yeah. That would do it. Thanks for reading through the logs so carefully. 

Now how to do it tastefully is the question

**Comment by hmaarrfk** [comment]
I can't find anything crazy that pytorch is doing....

I think setting the CMAKE_STAGING_PREFIX is overriding the CMAKE_INSTALL_PREFIX that pytroch sets:
https://cmake.org/cmake/help/latest/variable/CMAKE_STAGING_PREFIX.html

From my logs
```
  --   CMAKE_INSTALL_PREFIX  : /home/conda/feedstock_root/build_artifacts/debug_1727047313939/work/torch

```
but it still goes to the CMAKE_STAGING_PREFIX

**Comment by hmaarrfk** [comment]
So i did the following:

```
mkdir simple
cd simple
cat >CMakeLists.txt <<EOF
cmake_minimum_required(VERSION 3.15)
project(MyCProject C)
find_package(Python3 REQUIRED COMPONENTS Interpreter Development NumPy)
set(CMAKE_C_STANDARD 99)
include_directories(${Python3_INCLUDE_DIRS})
include_directories(${Python3_NumPy_INCLUDE_DIRS})
add_executable(my_executable main.c)
target_link_libraries(my_executable Python3::Python)
EOF

mkdir build
cd build
cmake ${CMAKE_ARGS} ..
```


And I still get the same error:
```
-- The C compiler identification is GNU 13.3.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /home/conda/feedstock_root/build_artifacts/debug_1727492314322/_build_env/bin/aarch64-conda-linux-gnu-cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
CMake Error at /home/conda/feedstock_root/build_artifacts/debug_1727492314322/_build_env/share/cmake-3.30/Modules/FindPackageHandleStandardArgs.cmake:233 (message):
  Could NOT find Python3 (missing: Python3_NumPy_INCLUDE_DIRS NumPy) (found
  version "3.12.6")
Call Stack (most recent call first):
  /home/conda/feedstock_root/build_artifacts/debug_1727492314322/_build_env/share/cmake-3.30/Modules/FindPackageHandleStandardArgs.cmake:603 (_FPHSA_FAILURE_MESSAGE)
  /home/conda/feedstock_root/build_artifacts/debug_1727492314322/_build_env/share/cmake-3.30/Modules/FindPython/Support.cmake:4001 (find_package_handle_standard_args)
  /home/conda/feedstock_root/build_artifacts/debug_1727492314322/_build_env/share/cmake-3.30/Modules/FindPython3.cmake:602 (include)
  CMakeLists.txt:5 (find_package)
```

**Comment by hmaarrfk** [comment]
to remove ambiguity that it is caused by the pytorch build system, i've simply created a simple CMakeLists.txt file.

File: recipe/build.sh [hunk 4]
```diff
@@ -2,10 +2,19 @@
 
 set -ex
 
+mkdir build
+cd build
+cmake ${CMAKE_ARGS} ..
```
[the remark this answers is not in the document]
Comment (traversaro) [reply recipe/build.sh]: I see, did you tried instead:

~~~
cmake ${CMAKE_ARGS} -DPython3_NumPy_INCLUDE_DIRS:PATH=$SP_DIR/numpy/_core/include -DPython_NumPy_INCLUDE_DIRS:PATH=$SP_DIR/numpy/_core/include  ..
~~~

?

That is what was working for me in https://github.com/conda-forge/idyntree-feedstock/pull/102 .

[the remark this answers is not in the document]
Comment (traversaro) [reply recipe/build.sh]: My bad, I did a common error, the correct variable to set is `Python3_NumPy_INCLUDE_DIR`, not `Python3_NumPy_INCLUDE_DIRS` **note the missing final S**, see https://cmake.org/cmake/help/latest/module/FindPython.html#artifacts-specification .

So I would try either:

~~~
cmake ${CMAKE_ARGS} -DPython3_NumPy_INCLUDE_DIR:PATH=$(python -c 'import numpy;print(numpy.get_include())') -DPython_NumPy_INCLUDE_DIR:PATH=$(python -c 'import numpy;print(numpy.get_include())')  ..
~~~

or 

~~~
cmake ${CMAKE_ARGS} -DPython3_NumPy_INCLUDE_DIR:PATH=$SP_DIR/numpy/_core/include -DPython_NumPy_INCLUDE_DIR:PATH=$SP_DIR/numpy/_core/include  ..
~~~

[the remark this answers is not in the document]
Comment (hmaarrfk) [reply recipe/build.sh]: I just think something has changed in the tooling.

You are correct, those variables USED to work.

I've tried many of them.

[the remark this answers is not in the document]
Comment (traversaro) [reply recipe/build.sh]: Not sure why, but in https://github.com/conda-forge/idyntree-feedstock/pull/109 it seems that those hints are still working somehow.

File: recipe/build.sh [hunk 5]
```diff
@@ -2,10 +2,29 @@
...
+Python3_NumPy_INCLUDE_DIR="$(python -c 'import numpy;print(numpy.get_include())')"
+Python3_NumPy_INCLUDE_DIRS="$(python -c 'import numpy;print(numpy.get_include())')"
+
+Python_NumPy_INCLUDE_DIR="$(python -c 'import numpy;print(numpy.get_include())')"
+Python_NumPy_INCLUDE_DIRS="$(python -c 'import numpy;print(numpy.get_include())')"
+cmake ${CMAKE_ARGS}                                                \
+    -DPython3_NumPy_INCLUDE_DIRS:PATH=DPython3_NumPy_INCLUDE_DIRS  \
+    -DPython3_NumPy_INCLUDE_DIR:PATH=DPython3_NumPy_INCLUDE_DIR    \
+    -DPython_NumPy_INCLUDE_DIRS:PATH=DPython_NumPy_INCLUDE_DIRS    \
+    -DPython_NumPy_INCLUDE_DIR:PATH=DPython_NumPy_INCLUDE_DIR      \
```
Comment (traversaro) [inline recipe/build.sh:16]:
```suggestion
    -DPython3_NumPy_INCLUDE_DIRS:PATH=$Python3_NumPy_INCLUDE_DIRS  \
    -DPython3_NumPy_INCLUDE_DIR:PATH=$Python3_NumPy_INCLUDE_DIR    \
    -DPython_NumPy_INCLUDE_DIRS:PATH=$Python_NumPy_INCLUDE_DIRS    \
    -DPython_NumPy_INCLUDE_DIR:PATH=$Python_NumPy_INCLUDE_DIR      \
```


Is this a typo or I am missing something?

Comment (hmaarrfk) [reply recipe/build.sh]: thanks.... sorry about  that

Comment (hmaarrfk) [reply recipe/build.sh]: wow you have revived my thoughts....

**Comment by Tobias-Fischer** [comment]
Wow seems like you finally got there @hmaarrfk, thanks and well done :)!

**Comment by hmaarrfk** [comment]
Ok lets take it one step at a time to stop the bleed.
https://github.com/conda-forge/pytorch-cpu-feedstock/pull/267

**Pull request closed without merging**